infrastructure/src/test/java/org/corfudb/infrastructure/CorfuSequencerServerTest.java

```diff
@@ -0,0 +1,528 @@
+        return requestHeader.getRequestId() == responseHeader.getRequestId() &&
+                requestHeader.getEpoch() == responseHeader.getEpoch() &&
+                requestHeader.getClientId().equals(responseHeader.getClientId()) &&
+                requestHeader.getClusterId().equals(responseHeader.getClusterId());
+    }
+
+    /**
+     * Perform the required preparation before running individual tests.
+     * This includes preparing the mocks and initializing the DirectExecutorService.
+     */
+    @Before
+    public void setup() {
+        // Initialize with newDirectExecutorService to execute the server RPC
+        // handler methods on the calling thread
+        when(mockServerContext.getExecutorService(anyInt(), anyString()))
+                .thenReturn(MoreExecutors.newDirectExecutorService());
+    }
+
+    /**
+     * Test that the SequencerServer sends NotReadyErrorMsg when client sends a TokenRequestMsg
+     * without bootstrapping it before.
+     * Essentially tests the {@link SequencerServer#isServerReadyToHandleMsg(RequestMsg)} method.
+     */
+    @Test
+    public void testSequencerServerNotReady() {
+        // When ServerContext epoch and sequencer epoch are not equal, it expects the bootstrap messages first.
+        // ServerContext epoch is set only by the bootstrap request. Make it return the Layout.INVALID_EPOCH.
+        when(mockServerContext.getServerEpoch()).thenReturn(Layout.INVALID_EPOCH);
+        sequencerServer = new SequencerServer(mockServerContext, spyFactoryHelper);
+        // Make sequencer
+        sequencerServer.setSequencerEpoch(0L);
+        RequestMsg request = getRequestMsg(
+                getBasicHeader(true, true),
+                CorfuProtocolSequencer.getTokenRequestMsg(0, Collections.emptyList())
+        );
+
+        ArgumentCaptor<ResponseMsg> responseCaptor = ArgumentCaptor.forClass(ResponseMsg.class);
+        sequencerServer.handleMessage(request, mockChannelHandlerContext, mockServerRouter);
+
+        verify(mockServerRouter).sendResponse(responseCaptor.capture(), any(ChannelHandlerContext.class));
+        ResponseMsg response = responseCaptor.getValue();
+
+        // Assert that the payload has a SERVER_ERROR and that the base
+        // header fields have remained the same
+        assertTrue(compareBaseHeaderFields(request.getHeader(), response.getHeader()));
+        assertTrue(response.getPayload().hasServerError());
+        assertTrue(response.getPayload().getServerError().hasNotReadyError());
+    }
+
+
+    /**
+     * Test that if sequencer server's epoch is Layout.INVALID_EPOCH (startup) OR
+     * if epoch sent by the client is not the consecutive epoch of the sequencerEpoch
+     * then the sequencer should not accept bootstrapWithoutTailsUpdate BootstrapSequencerRequestMsg,
+     * and return a Sequencer.BootstrapSequencerResponseMsg.NACK message.
+     */
+    @Test
+    public void testRequireFullBootstrap() {
```

Comment:
Same here. Consider also verifying that methods like `setSequencerEpoch` were not called.